deps/compile.jl

```diff
@@ -0,0 +1,182 @@
+
+    flags = [ "--compiler-bindir" ]
+
+    # collect possible hostcc executable names
+    if !is_windows()
+        # Determine host compiler version requirements; Source: "CUDA Getting Started Guide for Linux".
+        hostcc_support = (
+            (v"5.0", v"4.6.4"),
+            (v"5.5", v"4.7.3"),
+            (v"6.0", v"4.8.1"),
+            (v"6.5", v"4.8.2"),
+            (v"7.0", v"4.9.2"),
+            (v"7.5", v"4.9.2"),
+            (v"8.0", v"5.3.1") )
+        if version < hostcc_support[1][1]
+            error("no support for CUDA < $(hostcc_req[1][1])")
+        end
+        hostcc_maxver = Nullable{VersionNumber}()
+        for hostcc in hostcc_support
+            if version == hostcc[1]
+                hostcc_maxver = Nullable(hostcc[2])
+                break
+            end
+        end
+        isnull(hostcc_maxver) && error("Unknown NVIDIA CUDA compiler version $version.")
+
+        hostcc_names = [ "gcc" ]
+        for ver in (v"4.4", v"4.5", v"4.6", v"4.7", v"4.8", v"4.9")
+            push!(hostcc_names, "gcc-$(ver.major).$(ver.minor)")
+            push!(hostcc_names, "gcc$(ver.major)$(ver.minor)")
+        end
+
+        # Check availability host compiler
+        hostcc_possibilities = []
+        for hostcc in hostcc_names
+            hostcc_path = try
+                chomp(readstring(pipeline(`which $hostcc`, stderr=DevNull)))
+            catch ex
+                !isa(ex, ErrorException) && rethrow(ex)
+                continue
+            end
+
+            verstring = chomp(readlines(`$hostcc_path --version`)[1])
+            m = match(Regex("^$hostcc \\(.*\\) ([0-9.]+)"), verstring)
+            if m === nothing
+                warn("Could not parse GCC version info (\"$verstring\"), skipping this compiler.")
+                continue
+            end
+            hostcc_ver = VersionNumber(m.captures[1])
+
+            if hostcc_ver <= get(hostcc_maxver)
+                push!(hostcc_possibilities, (hostcc_path, hostcc_ver))
+            end
+        end
+        if length(hostcc_possibilities) == 0
+            error("Could not find a suitable host compiler (your NVCC $version needs GCC <= $(get(hostcc_maxver))).")
+        end
+        sort!(hostcc_possibilities; rev=true, lt=(a, b) -> a[2]<b[2])
+        hostcc = hostcc_possibilities[1]
+
+        push!(flags, hostcc[1])
+    else # Windows
+        vc_versions = ["VS140COMNTOOLS", "VS120COMNTOOLS", "VS110COMNTOOLS", "VS100COMNTOOLS"]
+        !any(x -> haskey(ENV, x), vc_versions) && error("Compatible Visual Studio installation cannot be found; Visual Studio 2015, 2013, 2012, or 2010 is required.")
+        vs_cmd_tools_dir = ENV[vc_versions[first(find(x -> haskey(ENV, x), vc_versions))]]
+        hostccbin = joinpath(dirname(dirname(dirname(vs_cmd_tools_dir))), "VC", "bin", Sys.WORD_SIZE == 64 ? "amd64" : "", "cl.exe")
+
+        push!(flags, hostccbin)
+    end
+
+    # Determine compilation options
+    if haskey(ENV, "ARCH")
```

Comment:
The environment variable should probably be either called `GPU_ARCH`, `DEVICE_ARCH`, or `NVCC_ARCH` so that it is not confused with `ARCH` which is used in some places (the Julia Makefile) as an example to denote the host architecture.